objects/src/accounts/storage/mod.rs

```diff
@@ -253,99 +312,89 @@ impl AccountStorage {
                     });
                 }
             },
-            slot_type => Err(AccountError::StorageSlotNotValueSlot(index, slot_type))?,
+            StorageSlotType::Map { value_arity } => {
+                if value_arity > 0 {
+                    return Err(AccountError::StorageSlotInvalidValueArity {
+                        slot: index,
+                        expected: 0,
+                        actual: value_arity,
+                    });
+                }
+            },
```

Comment:
I think this should be an error as we should not be able to set the root of a storage map directly.

Reply:
Not sure if I understand, if this would be an error, we could not update a map. The `AccountDelta` creates for every `updated_map` a corresponding update in `set_item` for the new map root. This is how we encountered the bug in the first place. When we try to update a `StorageSlot` of `Type` `Map`.

Reply:
But I don't think we call this method for updating storage maps. Storage maps are updated directly by either applying delta to them or via `set_map_item()` method. Or am I missing something?

Reply:
This time, you are missing something, I guess. 

StorageMaps are updated directly, but the corresponding `StorageMapRoot` is a `SlotItem` of `Type::Map` that is updated with every StorageMapUpdate.

You can remove this part and the test will fail.

```
    # set the new map item
    loc_load.0 exec.account::set_map_item swapw
    # => [NEW_MAP_ROOT, OLD_MAP_VALUE, ...]

    # set the root of the map in the respective account storage slot
    loc_load.0 exec.account::set_item
    # => [OLD_MAP_ROOT, OLD_MAP_VALUE, ...]
```

Here in `account.masm` you can see that "at the same time" we emit one event to set a map item and another to set the new root as slot item. 

That means if we remove that part 

```            StorageSlotType::Map { value_arity } => {
                if value_arity > 0 {
                    return Err(AccountError::StorageSlotInvalidValueArity {
                        slot: index,
                        expected: 0,
                        actual: value_arity,
                    });
                }
            },
```

We could not change the root of a map as SlotItem.

Reply:
After we merge this PR we should add a test in the client that changes a map @igamigo

Reply:
Hmmm - I may have a wrong model of how `apply_delta()` works. I thought `cleared_items` and `updated_items` contain updates only for value slots - but it seems like they contain updates for roots of other data structures as well?

If so, I'm not sure that's a good idea as it introduces a potential area of inconsistency (i.e., it is possible to construct a storage delta in a way which would put account storage into an invalid state after update).

If the above is correct, we should probably change how storage delta works as it related to storage maps. Specifically, all updates related to storage maps should be contained within the `AccountStorageDelta.updated_maps` field.

Reply:
I see, we have several checks that a StorageMap item always needs a corresponding SlotItem change. 

At least in MASM, we set the new `StorageMapRoot` when we get it using the `set_item` procedure. This triggers and update in `updated_items`. 

In my understanding, for every change in a map, there is always a change in the corresponding slot item that stores the map's root. 

I see three options:

1. We can not apply that change in MASM directly, but that would lead to an inconsistent AccountStorageState during the transaction. Then we could store all updates related to storage maps in `AccountStorageDelta.updated_maps`. We can then apply the delta and in doing so change the corresponding slot item. 
2. Or we could introduce a new procedure in MASM that changes the SlotItem but does not trigger an event. 
3. We could have some checks in `apply_delta` to prevent inconsistencies. So for every map change, we need a corresponding change in the SlotItem (this is what we do currently).

Reply:
I think the second option would be the ideal one. It will probably be quite a bit of work thought and this PR may not be the right place to do it.

So, we can merge this PR and then tackle storage-map related issues in a follow-up PR.

btw, I thought we added `AccountStorage::set_map_item()` method, but I can't find it now.

Reply:
Ah I moved the logic to `AccountStorage::apply_delta`

Reply:
I think we should have both: `set_storage_item()` which is similar to the corresponding procedure in MASM, and `apply_delta()`.

Reply:
I added the `set_map_item` method. 

The signature is not equal to the MASM procedure since we are lacking the `index` in `map_updates`.